jax/lax/lax_parallel.py

```diff
@@ -557,13 +557,25 @@ def _all_to_all_translation_rule(c, x, *, split_axis, concat_axis, axis_name,
   replica_groups = _replica_groups(axis_env, axis_name, None)
   if len(replica_groups[0]) == 1:
     return x
-  else:
+  elif platform == 'tpu':
     split_count = len(replica_groups[0])
     if not all(split_count == len(g) for g in replica_groups):
       raise ValueError('Replica groups must be equally sized')
     replica_groups_protos = xc.make_replica_groups(replica_groups)
-    return xops.AllToAll(x, split_axis, concat_axis, split_count,
-                         replica_groups_protos)
+    if concat_axis == split_axis:
+      return xops.AllToAll(x, split_axis, concat_axis, split_count,
+                           replica_groups_protos)
+    else:
+      if concat_axis < split_axis:
+        split_axis += 1
+      elif split_axis < concat_axis:
+        concat_axis += 1
+      x = xla.lower_fun(partial(lax.expand_dims, dimensions=(concat_axis,)), multiple_results=False)(c, x)
+      x = xops.AllToAll(x, split_axis, concat_axis, split_count, replica_groups_protos)
+      x = xla.lower_fun(partial(lax.squeeze, dimensions=(split_axis,)), multiple_results=False)(c, x)
```

Comment:
The semantics of XLA's `AllToAll` differ from the JAX's interpretation in that in XLA the collective doesn't insert or remove any dimensions, but only moves chunks of data from one existing dimension to the other existing dimension. Here, I simulate our behavior by inserting the concat axis, running all_to_all, and then squeezing out the now singleton split axis.

Reply:
I admire your laziness in using `xla.lower_fun` here rather than writing the reshapes in terms of the HLO builder :P

Reply:
🙃

Reply:
I'm still trying to put off the day in which I'll have to learn more XLA operators.